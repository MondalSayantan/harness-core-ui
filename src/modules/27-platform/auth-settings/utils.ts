/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get } from 'lodash-es'
import type { CreateIpAllowlistConfigProps, UpdateIpAllowlistConfigProps } from '@harnessio/react-ng-manager-client'

import type { IIPAllowlistForm } from '@auth-settings/interfaces/IPAllowlistInterface'
import { SourceType } from '@auth-settings/interfaces/IPAllowlistInterface'
import { CurrencyType } from '@common/constants/SubscriptionTypes'

export function getAmountInCurrency(currency: CurrencyType, amount: number): string {
  switch (currency) {
    case CurrencyType.USD: {
      return amount.toLocaleString('en-US', {
        style: 'currency',
        currency: 'USD'
      })
    }
  }

  return ''
}

// dividing the actual amount by 100 as amount numbers are in cents
export function getDollarAmount(amount?: number, isYearly?: boolean): number {
  if (isYearly) {
    return (amount || 0) / 1200
  }
  return (amount || 0) / 100
}

export const getErrorMessage = (error: any): string =>
  get(error, 'data.error', get(error, 'data.message', error?.message))

export const buildCreateIPAllowlistPayload = (data: IIPAllowlistForm): CreateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: false, // always false while creating
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags
      }
    }
  }
}

export const buildUpdateIPAllowlistPayload = (
  data: IIPAllowlistForm,
  overrideFields?: any
): UpdateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    'ip-config-identifier': data.identifier,
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: data?.enabled,
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags,
        ...overrideFields
      }
    }
  }
}

/**
 * Converts minutes to higher dimensions of time such as months, days and hours
 *
 * Examples:
 * formatMinutesToHigherDimensions(150);    // Output: "2 hours 30 minutes"
 * formatMinutesToHigherDimensions(4320);   // Output: "3 days"
 * formatMinutesToHigherDimensions(45000);  // Output: "1 month 1 day 6 hours"
 * @param minutes
 * @returns a string in higher dimensions
 */
export const formatMinutesToHigherDimensions = (minutes: number | undefined): string => {
  if (minutes === undefined || minutes < 0) {
    return 'Invalid input'
  }

  const MINUTES_IN_AN_HOUR = 60
  const HOURS_IN_A_DAY = 24
  const DAYS_IN_A_MONTH = 30

  const result = []

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH) {
    const months = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH))
    result.push(`${months} month${months > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY * DAYS_IN_A_MONTH
  }

  if (minutes >= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY) {
    const days = Math.floor(minutes / (MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY))
    result.push(`${days} day${days > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR * HOURS_IN_A_DAY
  }

  if (minutes >= MINUTES_IN_AN_HOUR) {
    const hours = Math.floor(minutes / MINUTES_IN_AN_HOUR)
    result.push(`${hours} hour${hours > 1 ? 's' : ''}`)
    minutes %= MINUTES_IN_AN_HOUR
  }

  if (minutes > 0) {
    result.push(`${minutes} minute${minutes > 1 ? 's' : ''}`)
  }

  return result.join(' ')
}
