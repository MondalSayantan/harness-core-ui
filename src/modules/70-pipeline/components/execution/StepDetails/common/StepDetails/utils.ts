/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { ResponseApprovalInstanceResponse } from 'services/pipeline-ng'

export function encodeURIWithReservedChars(uri: string): string {
  return encodeURIComponent(uri).replace(/[;,/?:@&=+$#]/g, function (c) {
    return '%' + c.charCodeAt(0).toString(16)
  })
}

export interface ExecutionMetadataType {
  approvalInstanceId: string
  mock?: {
    data?: ResponseApprovalInstanceResponse
    loading?: boolean
  }
}
