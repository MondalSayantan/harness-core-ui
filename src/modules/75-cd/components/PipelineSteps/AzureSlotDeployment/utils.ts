import { RUNTIME_INPUT_VALUE, MultiTypeInputType, getMultiTypeFromValue } from '@harness/uicore'
import type { StageElementWrapper } from '@pipeline/utils/pipelineTypes'
import type { StageElementConfig, StageElementWrapperConfig, PipelineInfoConfig } from 'services/pipeline-ng'
import type { InfraStructureDefinitionYaml, DeploymentStageConfig, EnvironmentYamlV2 } from 'services/cd-ng'

interface EnvironmentYamlV2Runtime extends Omit<EnvironmentYamlV2, 'infrastructureDefinitions'> {
  infrastructureDefinitions?: InfraStructureDefinitionYaml[] | string
}

interface DeploymentStageConfigAzureWebName extends Omit<DeploymentStageConfig, 'environment'> {
  environment: EnvironmentYamlV2Runtime
}

export interface DeploymentStageElementConfigAzureWebName extends Omit<StageElementConfig, 'spec'> {
  spec?: DeploymentStageConfigAzureWebName
}
export type SelectedStageType = StageElementWrapper<DeploymentStageElementConfigAzureWebName>

export function getStage(stageId: string, pipeline?: PipelineInfoConfig): StageElementWrapperConfig | undefined {
  if (pipeline?.stages) {
    let responseStage: StageElementWrapperConfig | undefined = undefined
    pipeline.stages.forEach(item => {
      if (item.stage && item.stage.identifier === stageId) {
        responseStage = item
      } else if (item.parallel) {
        return item.parallel.forEach(node => {
          if (node.stage?.identifier === stageId) {
            responseStage = node
          }
        })
      }
    })
    return responseStage
  }
  return
}

export const isRuntimeEnvId = (selectedStage: SelectedStageType): boolean => {
  const envId = selectedStage?.stage?.spec?.environment?.environmentRef
  return !!(envId && envId === RUNTIME_INPUT_VALUE)
}

export const isRuntimeInfraId = (selectedStage: SelectedStageType): boolean => {
  const infraId = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions
  return infraId === RUNTIME_INPUT_VALUE
}

export const isMultiEnv = (selectedStage: SelectedStageType): boolean => {
  const isEnvs = !!selectedStage?.stage?.spec?.environments || !!selectedStage?.stage?.spec?.environmentGroup
  return !!isEnvs
}

export const getEnvId = (selectedStage: SelectedStageType): string => {
  if (isMultiEnv(selectedStage || isRuntimeEnvId(selectedStage))) {
    return ''
  }

  return (
    selectedStage?.stage?.spec?.environment?.environmentRef ||
    selectedStage?.stage?.template?.templateInputs?.spec?.environment?.environmentRef ||
    ''
  )
}

export const getInfraId = (selectedStage: SelectedStageType, infraSourceField?: string): string => {
  const envId = getEnvId(selectedStage)
  if (!envId || isMultiEnv(selectedStage)) {
    return ''
  }

  if (!isRuntimeEnvId(selectedStage) && !isRuntimeInfraId(selectedStage)) {
    const [infra] = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField) {
      return infra?.inputs?.spec?.[infraSourceField]
    }
    return infra.identifier
  }
  return ''
}

/* eslint-disable */
export const getEnvIdRuntime = (stageId: string, values: any): string => {
  const stage = getStage(stageId, values) as SelectedStageType
  const envId =
    stage?.stage?.spec?.environment?.environmentRef ||
    stage?.stage?.template?.templateInputs?.spec?.environment?.environmentRef
  return envId || ''
}
/* eslint-disable */
export const getInfraIdRuntime = (stageId: string, values: any, infraSourceField?: string): string => {
  const stage = getStage(stageId, values) as SelectedStageType
  const isTemplateStageInfra = stage?.stage?.template?.templateInputs?.spec?.environment?.infrastructureDefinitions?.[0]
  if (isTemplateStageInfra) {
    const [{ inputs, identifier }] = stage?.stage?.template?.templateInputs?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField && infraSourceField === 'infrasctructure') {
      return identifier
    }
    if (infraSourceField && inputs?.spec) {
      return inputs?.spec?.[infraSourceField]
    }
  }

  if (stage?.stage?.spec?.environment?.infrastructureDefinitions?.[0]) {
    const [{ identifier, inputs }] = stage?.stage?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    if (infraSourceField && inputs?.spec && infraSourceField !== 'infrasctructure') {
      return inputs?.spec?.[infraSourceField]
    }

    return identifier
  }
  return ''
}

export const getAllowableTypes = (selectedStage: SelectedStageType): MultiTypeInputType[] => {
  if (isMultiEnv(selectedStage)) {
    return [MultiTypeInputType.EXPRESSION]
  }

  return [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]
}

export const infrastructurePath = 'template.templateInputs.spec.environment.infrastructureDefinitions[0]'
export const resourceGroupPath = `${infrastructurePath}.inputs.spec.resourceGroup`
export const connectorPath = `${infrastructurePath}.inputs.spec.connectorRef`
export const subscriptionPath = `${infrastructurePath}.inputs.spec.subscriptionId`

export const getInfraParamsFixedValue = (values: string[]) => {
  let fixedValue = ''
  values?.forEach((value: string) => {
    if (value && getMultiTypeFromValue(value) === MultiTypeInputType.FIXED) {
      fixedValue = value
    }
  })
  return fixedValue
}
