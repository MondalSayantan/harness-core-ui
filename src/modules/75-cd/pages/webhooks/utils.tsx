/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Scope } from '@common/interfaces/SecretsInterface'

export interface AddWebhookModalData {
  name: string
  identifier: string
  connectorRef: string
  repo: string
  folderPaths: { id: string; value: string }[] | string
}

export interface NewWebhookModalProps {
  isEdit: boolean
  initialData: AddWebhookModalData
  entityScope?: Scope
  closeModal: () => void
}

export const initialWebhookModalData = {
  name: '',
  identifier: '',
  connectorRef: '',
  repo: '',
  folderPaths: []
}
