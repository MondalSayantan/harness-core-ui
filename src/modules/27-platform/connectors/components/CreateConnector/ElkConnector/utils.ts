/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import { UseStringsReturn } from 'framework/strings'
import type { ConnectorConfigDTO } from 'services/cd-ng'
import { ElkConnectorAuthFieldsProps } from './CreateElkConnector.types'
import { ELKConnectorFields, ElkAuthType } from './CreateElkConnector.constants'

type InitializeElkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeElkConnector({
  prevStepData,
  accountId,
  projectIdentifier,
  orgIdentifier
}: InitializeElkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    accountId,
    username: '',
    authType: AuthTypes.USER_PASSWORD,
    password: undefined,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }
  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.controllerUrl || spec?.url || '',
    password: prevData?.password || spec?.passwordRef,
    clientSecretRef: prevData?.clientSecretRef || spec?.clientSecretRef,
    authType: prevData?.authType || spec?.authType || AuthTypes.USER_PASSWORD,
    username: prevData?.username || spec?.username || '',
    clientId: prevData?.clientId || spec?.clientId,
    apiKeyId: prevData?.apiKeyId || spec?.apiKeyId,
    apiKeyRef: prevData?.apiKeyRef || spec?.apiKeyRef
  }
}

interface GetElkAuthTypeReturn {
  label: string
  value: string
}

export const getELKAuthType = (getString: UseStringsReturn['getString']): GetElkAuthTypeReturn[] => [
  {
    label: getString('usernamePassword'),
    value: AuthTypes.USER_PASSWORD
  },
  {
    label: getString('common.apikey'),
    value: AuthTypes.API_CLIENT_TOKEN
  },
  {
    label: getString('platform.connectors.bearerToken'),
    value: AuthTypes.BEARER_TOKEN
  },
  {
    label: getString('platform.connectors.elk.noAuthentication'),
    value: ElkAuthType.NONE
  }
]

export const getAuthFields = (
  getString: UseStringsReturn['getString'],
  authType?: string
): ElkConnectorAuthFieldsProps[] | null => {
  if (!authType) {
    return null
  }

  switch (authType) {
    case AuthTypes.USER_PASSWORD:
      return [
        { name: ELKConnectorFields.USERNAME, label: getString('username'), key: ELKConnectorFields.USERNAME },
        { name: ELKConnectorFields.PASSWORD, label: getString('password'), key: ELKConnectorFields.PASSWORD }
      ]

    case AuthTypes.API_CLIENT_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_ID,
          label: getString('platform.connectors.elk.apiId'),
          key: ELKConnectorFields.API_KEY_ID
        },
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('common.apikey'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    case AuthTypes.BEARER_TOKEN:
      return [
        {
          name: ELKConnectorFields.API_KEY_REF,
          label: getString('platform.connectors.bearerToken'),
          key: ELKConnectorFields.API_KEY_REF
        }
      ]

    default:
      return null
  }
}
