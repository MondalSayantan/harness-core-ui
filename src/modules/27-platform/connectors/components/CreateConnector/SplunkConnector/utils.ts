/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormData } from '@platform/connectors/interfaces/ConnectorInterface'
import { AuthTypes } from '@platform/connectors/pages/connectors/utils/ConnectorHelper'
import type { ConnectorConfigDTO } from 'services/cd-ng'

type InitializeSplunkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeSplunkConnector({
  prevStepData,
  projectIdentifier,
  accountId,
  orgIdentifier
}: InitializeSplunkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    authType: AuthTypes.USER_PASSWORD,
    username: '',
    passwordRef: undefined,
    tokenRef: undefined,
    accountId: accountId,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }

  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.splunkUrl || '',
    username: prevData?.username || spec?.username,
    passwordRef: prevData?.passwordRef || spec?.passwordRef,
    authType: prevData?.authType || spec?.type || defaultObj.authType,
    tokenRef: prevData?.tokenRef || spec?.tokenRef
  }
}
