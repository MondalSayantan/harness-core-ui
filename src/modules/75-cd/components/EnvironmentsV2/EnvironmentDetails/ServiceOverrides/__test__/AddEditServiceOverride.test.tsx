/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as cdNgServices from 'services/cd-ng'
import routes from '@common/RouteDefinitions'
import { environmentPathProps, projectPathProps } from '@common/utils/routeUtils'
import { TestWrapper } from '@common/utils/testUtils'
import AddEditServiceOverride from '../AddEditServiceOverride'
import mockServiceList from './__mocks__/mockServicesListForOverrides.json'

jest.mock('services/cd-ng', () => ({
  ...jest.requireActual('services/cd-ng'),
  useUpsertServiceOverride: jest.fn().mockImplementation(() => {
    return {
      loading: false,
      mutate: jest.fn().mockResolvedValue({}),
      cancel: jest.fn(),
      error: null
    }
  }),
  useGetService: jest.fn().mockImplementation(() => ({ loading: false, data: {}, refetch: jest.fn() })),
  getServiceAccessListPromise: jest.fn().mockImplementation(() => Promise.resolve(mockServiceList))
}))

describe('Add Edit Service Override Test', () => {
  test('add new service override', async () => {
    jest.spyOn(cdNgServices, 'useUpsertServiceOverride')

    const { container, getByTestId } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'test_env'
        })}
        pathParams={{ ...projectPathProps, ...environmentPathProps }}
      >
        <AddEditServiceOverride
          defaultTab="variableoverride"
          closeModal={jest.fn()}
          expressions={['']}
          selectedService={null}
          isReadonly={false}
          services={[]}
        />
      </TestWrapper>
    )

    await userEvent.click(getByTestId('cr-field-serviceRef') as HTMLElement)

    const serviceText = screen.getByText('custom test')
    expect(serviceText).toBeInTheDocument()

    await userEvent.click(serviceText)
    await userEvent.click(screen.getByText('entityReference.apply')!)
    await userEvent.click(screen.getByText('variableLabel'))

    const newOverrideText = screen.getByText('common.newName common.override')
    expect(newOverrideText).toBeDefined()
    await userEvent.click(newOverrideText)

    // select variable
    const selectBoxes = screen.getAllByPlaceholderText('- common.selectName -')
    const variableTextBox = selectBoxes[0]
    await userEvent.type(variableTextBox, 'var2')
    await userEvent.click(container.querySelector('[data-icon="plus"]') as HTMLElement)
    await waitFor(() => {
      expect(screen.queryByText('var2')).toBeInTheDocument()
    })

    await waitFor(async () => {
      const inputs = (await screen.queryAllByRole('textbox')) as HTMLElement[]
      const overrideValueInput = inputs[inputs.length - 1]

      expect(overrideValueInput).toBeDefined()
      await fireEvent.change(overrideValueInput, {
        target: { value: 'testStr' }
      })
    })

    await act(async () => {
      userEvent.click(screen.getByText('submit'))
    })

    await userEvent.click(screen.getAllByRole('button')[1])
    await waitFor(() => {
      expect(screen.getAllByRole('button')[1]).toHaveClass('PillToggle--item PillToggle--selected')
    })

    await userEvent.click(screen.getAllByRole('button')[0])
    await waitFor(() => {
      expect(screen.getAllByRole('button')[0]).toHaveClass('PillToggle--item PillToggle--selected')
    })

    expect(container).toMatchSnapshot()
  })
})
