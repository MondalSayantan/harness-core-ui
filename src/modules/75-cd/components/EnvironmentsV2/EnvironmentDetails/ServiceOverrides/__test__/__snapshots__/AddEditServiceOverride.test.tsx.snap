// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Edit Service Override Test add new service override 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <div
      class="serviceOverrideDialog"
    >
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--padding-bottom-medium"
      >
        <div
          class="PillToggle--optionBtns"
        >
          <div
            class="PillToggle--item PillToggle--selected"
            data-name="toggle-option-one"
            role="button"
            tabindex="0"
          >
            VISUAL
          </div>
          <div
            class="PillToggle--item"
            data-name="toggle-option-two"
            role="button"
            tabindex="0"
          >
            YAML
          </div>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <div
          class="serviceLabel"
        >
          <div
            class="StyledProps--font StyledProps--main"
            data-testid="serviceTooltip"
          >
            <span
              class="Tooltip--acenter bp3-label"
              data-tooltip-id="addEditServiceOverrideForm_serviceRef"
            >
              service
            </span>
          </div>
          <div
            class="bp3-form-group"
          >
            <div
              class="bp3-form-content"
            >
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main container Button--iconOnly Button--without-shadow Button--withRightIcon"
                data-testid="cr-field-serviceRef"
                style="width: 300px;"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  <div
                    class="Layout--horizontal Layout--layout-spacing-small StyledProps--main selectWrapper StyledProps--flex StyledProps--flex-distribution-space-between"
                  >
                    <span
                      class="bp3-popover-wrapper"
                    >
                      <span
                        class="bp3-popover-target Text--targetClass"
                      >
                        <p
                          class="StyledProps--font StyledProps--main label StyledProps--color StyledProps--color-grey800"
                          tabindex="0"
                        >
                          custom_test
                        </p>
                      </span>
                    </span>
                    <span
                      class="Tag--main"
                    >
                      <span
                        class="bp3-tag bp3-minimal"
                        id="tag"
                      >
                        <span
                          class="bp3-text-overflow-ellipsis bp3-fill"
                        >
                          account
                        </span>
                      </span>
                    </span>
                  </div>
                </span>
                <span
                  class="bp3-icon bp3-icon-chevron-down StyledProps--main"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="14"
                    viewBox="0 0 16 16"
                    width="14"
                  >
                    <desc>
                      chevron-down
                    </desc>
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <p
          class="StyledProps--font StyledProps--main"
        >
          common.serviceOverrides.overrideType
        </p>
        <div
          class="bp3-tabs Tabs--main Tabs--horizontal"
        >
          <div
            class="bp3-tab-list"
            role="tablist"
          >
            <div
              class="bp3-tab-indicator-wrapper"
              style="height: 0px; transform: translateX(0px) translateY(0px); width: 0px;"
            >
              <div
                class="bp3-tab-indicator"
              />
            </div>
            <div
              aria-controls="bp3-tab-panel_serviceOverrideTab_variableoverride"
              aria-disabled="false"
              aria-expanded="true"
              aria-selected="true"
              class="bp3-tab"
              data-tab-id="variableoverride"
              id="bp3-tab-title_serviceOverrideTab_variableoverride"
              role="tab"
              tabindex="0"
            >
              variableLabel
            </div>
            <div
              aria-controls="bp3-tab-panel_serviceOverrideTab_manifestoverride"
              aria-disabled="false"
              aria-expanded="false"
              aria-selected="false"
              class="bp3-tab"
              data-tab-id="manifestoverride"
              id="bp3-tab-title_serviceOverrideTab_manifestoverride"
              role="tab"
              tabindex="0"
            >
              manifestsText
            </div>
            <div
              aria-controls="bp3-tab-panel_serviceOverrideTab_configoverride"
              aria-disabled="false"
              aria-expanded="false"
              aria-selected="false"
              class="bp3-tab"
              data-tab-id="configoverride"
              id="bp3-tab-title_serviceOverrideTab_configoverride"
              role="tab"
              tabindex="0"
            >
              cd.configFileStoreTitle
            </div>
          </div>
          <div
            aria-hidden="false"
            aria-labelledby="bp3-tab-title_serviceOverrideTab_variableoverride"
            class="bp3-tab-panel"
            id="bp3-tab-panel_serviceOverrideTab_variableoverride"
            role="tabpanel"
          >
            <div
              class="Layout--vertical Layout--layout-spacing-medium StyledProps--main StyledProps--flex StyledProps--flex-alignItems-flex-start"
            >
              <div
                class="Layout--vertical StyledProps--main"
              >
                <div
                  class="tableRow headerRow"
                >
                  <p
                    class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
                  >
                    cd.configurationVariable
                  </p>
                  <p
                    class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
                  >
                    typeLabel
                  </p>
                  <p
                    class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
                  >
                    cd.overrideValue
                  </p>
                </div>
                <div
                  class="tableRow rowValue"
                >
                  <p
                    class="StyledProps--font Text--lineclamp StyledProps--main"
                    style="--text-line-clamp: 1;"
                  >
                    
                  </p>
                  <p
                    class="StyledProps--font Text--lineclamp StyledProps--main"
                    style="--text-line-clamp: 1;"
                  >
                    String
                  </p>
                  <p
                    class="StyledProps--font Text--lineclamp StyledProps--main"
                    style="--text-line-clamp: 1;"
                  >
                    testStr
                  </p>
                  <span>
                    <div
                      class="Layout--horizontal StyledProps--main variableListButtons"
                    >
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                            data-testid="edit-variable-0"
                            tabindex="0"
                            type="button"
                          >
                            <span
                              class="bp3-icon StyledProps--main"
                              data-icon="Edit"
                            >
                              <svg
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M11.207.485a.5.5 0 01.354-.146 3.5 3.5 0 013.5 3.5.5.5 0 01-.146.353l-11 11a.5.5 0 01-.354.147h-3a.5.5 0 01-.5-.5v-3a.5.5 0 01.146-.354l11-11zm.553.862L1.061 12.046v2.293h2.293l10.7-10.7a2.5 2.5 0 00-2.294-2.292z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                />
                                <path
                                  clip-rule="evenodd"
                                  d="M11.89 2.646a.5.5 0 010 .708L8.353 6.889a.5.5 0 01-.708-.707l3.536-3.536a.5.5 0 01.707 0z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  opacity="0.3"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <button
                            class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                            data-testid="delete-variable-0"
                            tabindex="0"
                            type="button"
                          >
                            <span
                              class="bp3-icon StyledProps--main"
                              data-icon="main-trash"
                            >
                              <svg
                                height="16"
                                viewBox="0 0 16 20"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                />
                                <path
                                  clip-rule="evenodd"
                                  d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  opacity="0.3"
                                />
                              </svg>
                            </span>
                          </button>
                        </span>
                      </span>
                    </div>
                  </span>
                </div>
              </div>
              <button
                aria-label="common.newName common.override"
                class="bp3-button Button--button StyledProps--font StyledProps--main addOverrideBtn Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link Button--size Button--small"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
                  icon="plus"
                >
                  <svg
                    data-icon="plus"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      plus
                    </desc>
                    <path
                      d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp3-button-text"
                >
                  common.newName common.override
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main StyledProps--padding-top-xxlarge"
      >
        <button
          aria-label="save"
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-primary"
          data-testid="addVariableSave"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            save
          </span>
        </button>
        <button
          aria-label="cancel"
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-tertiary"
          data-testid="addVariableCancel"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            cancel
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
